backends/p4tools/testgen/targets/bmv2/expr_stepper.cpp

```diff
@@ -585,13 +585,77 @@ void BMv2_V1ModelExprStepper::evalExternMethodCall(const IR::MethodCallExpressio
-             ::warning("counter.count not fully implemented.");
+         [this](const IR::MethodCallExpression* call, const IR::Expression* receiver,
+                IR::ID& /*methodName*/, const IR::Vector<IR::Argument>* args,
+                const ExecutionState& state, SmallStepEvaluator::Result& result) {
+             const auto* arg = args->at(0);
+             const auto* index = arg->expression;
+             if (state.hasTaint(index)) {
+                 ::warning(
+                     "Count index arg are tainted and not predictable. Skipping count execution.");
+                 auto* nextState = new ExecutionState(state);
+                 nextState->popBody();
+                 result->emplace_back(nextState);
+                 return;
+             }
+             // TODO: Frontload this in the expression stepper for method call expressions.
+             if (!SymbolicEnv::isSymbolicValue(index)) {
+                 // Evaluate the condition.
+                 stepToSubexpr(index, result, state, [call](const Continuation::Parameter* v) {
+                     auto* clonedCall = call->clone();
+                     auto* arguments = clonedCall->arguments->clone();
+                     auto* arg = arguments->at(0)->clone();
+                     arg->expression = v->param;
+                     (*arguments)[0] = arg;
+                     clonedCall->arguments = arguments;
+                     return Continuation::Return(clonedCall);
+                 });
+                 return;
+             }
+
+             const auto* receiverPath = receiver->checkedTo<IR::PathExpression>();
+             const auto& externInstance = state.convertPathExpr(receiverPath);
+             auto counterSizeExpr = state.findDecl(receiverPath)
+                                        ->checkedTo<IR::Declaration_Instance>()
+                                        ->arguments[0]
+                                        .at(0)
+                                        ->expression;
+             auto cond = new IR::Lss(index, counterSizeExpr);
              auto* nextState = new ExecutionState(state);
              nextState->popBody();
-             result->emplace_back(nextState);
+             // Retrieve the counter state from the object store. If it is already present, just
+             // cast the object to the correct class and retrieve the current value according to the
+             // index. If the counter has not been added had, create a new counter object.
+             // "Count" increment an entry by one. If index >= size, no counter state will be
+             // updated.
+             const auto* counterState =
+                 state.getTestObject("countervalues", externInstance->toString(), false);
+
+             Bmv2RegisterValue* counterValue = nullptr;
+             if (counterState != nullptr) {
+                 counterValue =
+                     new Bmv2RegisterValue(*counterState->checkedTo<Bmv2RegisterValue>());
```

Comment:
If you use a register object, you will have no means to differentiate them in the test back end. There is a lot of code overlap, but you should be able to tell when you have a counter or a register.